File: experiments/board-detection-2d/src/boarddet/square_fit.py
"""Fixed-size square model fit: refine (or rescue) a candidate's pose by
fitting a KNOWN-side square's 3 DOF (center, theta) directly to its 2D
points, instead of trusting the 2D quad's (raw-point minAreaRect) angle.

Why this exists (see .superpowers/sdd/stage7-stance-cause.md): on sparse
frames the quad's angle is essentially uncorrelated with truth (median error
43 deg), which is what makes the stance gate wrongly reject an upright board.
A square fit that KNOWS the true side length ahead of time doesn't need to
infer size from the data at all, so it can spend its degrees of freedom
purely on getting theta right -- median error drops to ~7 deg.

Deliberately NOT filled-square ICP: an ICP-style fit that only pulls its
model's occupied interior onto matched points has zero residual gradient
whenever the model already encloses every point (an "enclosing init" is a
fixed point -- interior points contribute nothing, since they're already
"inside"). The coverage residual below charges for BOTH (a) points that fall
outside the fixed-size square (it shouldn't be smaller than the data implies)
and (b) perimeter band it doesn't reach (it shouldn't be larger than the data
implies, either) -- so an over-large or mis-rotated enclosing box is still
penalized and the search has a gradient to follow.
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

_MIN_POINTS = 20
# Trim this much off each end when reading the points' own extent per axis
# (min/max would work on clean data; a couple of percent of headroom guards
# against a single stray outlier setting the whole box).
_EXTENT_PCTL = 2.0
_N_BINS_PER_SIDE = 10   # perimeter coverage-band bins per side (matches the
                        # stance-cause.md throwaway diagnostic's own choice)
_BAND_FRAC = 0.06       # coverage-band half-width, as a fraction of `side`
_COARSE_STEPS = 37      # steps across the full [0, 90) deg sweep (~2.5 deg)
_COARSE_REFINE_WINDOW_DEG = 4.0  # window for the post-coarse-sweep refine
_RESOLUTION_DEG = 0.25  # target step spacing for any windowed theta search
_LOCALIZE_RADIUS_FACTOR = 1.5  # x `side`, around init_center


@dataclass
class SquareFit:
    center: np.ndarray      # (2,) plane coords, same frame as coords_2d
    theta: float            # radians; determined mod 90 deg (4-fold symmetry)
    residual: float         # coverage residual, lower is better; ~0 is exact
    corners_2d: np.ndarray  # (4,2), cyclic order (not necessarily CCW)


def _rot2d(theta: float) -> np.ndarray:
    c, s = np.cos(theta), np.sin(theta)
    return np.array([[c, -s], [s, c]])


def _coverage_residual(rel: np.ndarray, side: float) -> float:
    """`rel` is points in the square's own axis-aligned frame, already
    centered on the candidate center (so the square occupies
    [-side/2, side/2]^2). Lower is better; 0 is an exact, fully-covered fit.
    """
    half = side / 2.0
    # Term 1: mean per-point distance outside the square. A real board's
    # points should all lie ON the board, so any point beyond the boundary
    # is a direct sign the box is mis-placed/rotated (this is what gives the
    # search a gradient even when the box already encloses everything --
    # see the module docstring).
    d_out = np.hypot(np.maximum(np.abs(rel[:, 0]) - half, 0.0),
                     np.maximum(np.abs(rel[:, 1]) - half, 0.0))
    mean_outside = float(d_out.mean())

    # Term 2: fraction of the perimeter's binned bands with no nearby point.
    # A real board's edge is a physical boundary scanned along its whole
    # length, so a correctly-posed box should have points near ~all of its
    # own perimeter, not just wherever the data happens to be densest.
    corners_sq = half * np.array([[1.0, 1.0], [-1.0, 1.0],
                                  [-1.0, -1.0], [1.0, -1.0]])
    band = _BAND_FRAC * side
    misses = 0
    for i in range(4):
        a, b = corners_sq[i], corners_sq[(i + 1) % 4]
        ab = b - a
        length = float(np.linalg.norm(ab))
        ab_hat = ab / length
        r = rel - a
        t = (r @ ab_hat) / length
        perp = np.abs(r[:, 0] * ab_hat[1] - r[:, 1] * ab_hat[0])
        near = (perp < band) & (t > -0.02) & (t < 1.02)
        bin_idx = np.clip((t[near] * _N_BINS_PER_SIDE).astype(np.int64), 0,
                          _N_BINS_PER_SIDE - 1)
        covered = len(np.unique(bin_idx))
        misses += _N_BINS_PER_SIDE - covered
    coverage_penalty = misses / (4 * _N_BINS_PER_SIDE)

    return mean_outside / side + coverage_penalty


def _fit_at_theta(coords_2d: np.ndarray, side: float, theta: float):
    """Closed-form center (robust per-axis extent midpoint) + coverage
    residual for a fixed theta. Returns (residual, center_world, corners_world).
    """
    rot = _rot2d(theta)
    p_sq = coords_2d @ rot  # world -> square-local (unrotate by theta)
    lo = np.percentile(p_sq, _EXTENT_PCTL, axis=0)
    hi = np.percentile(p_sq, 100.0 - _EXTENT_PCTL, axis=0)
    center_sq = (lo + hi) / 2.0
    rel = p_sq - center_sq

    residual = _coverage_residual(rel, side)

    half = side / 2.0
    corners_sq = half * np.array([[1.0, 1.0], [-1.0, 1.0],
                                  [-1.0, -1.0], [1.0, -1.0]])
    center_world = center_sq @ rot.T
    corners_world = (corners_sq + center_sq) @ rot.T
    return residual, center_world, corners_world


def _best_over(coords_2d: np.ndarray, side: float, thetas: np.ndarray):
    best_residual = np.inf
    best = None
    for theta in thetas:
        residual, center, corners = _fit_at_theta(coords_2d, side, theta)
        if residual < best_residual:
            best_residual = residual
            best = (theta, center, corners)
    theta, center, corners = best
    return SquareFit(center=center, theta=float(theta),
                     residual=float(best_residual), corners_2d=corners)


def fit_fixed_square(coords_2d: np.ndarray, side: float,
                     init_center: np.ndarray | None = None,
                     init_theta: float | None = None,
                     theta_window_deg: float = 20.0) -> SquareFit | None:
    """Fit a fixed-side-`side` square's center + theta to `coords_2d`.

    `init_theta=None`: no orientation prior at all -- coarse full sweep over
    [0, 90 deg) (4-fold symmetry: a square looks the same every 90 deg, so
    that range is the whole period), then a finer windowed refine around the
    coarse winner. Used both directly (dense-fixture tests) and as the
    fallback inside `init_theta`-less callers.

    `init_theta` given: search is a bounded window (+/- theta_window_deg)
    around it instead of the full range -- this is the normal "refine the
    quad's angle" call, and is what actually recovers pose on sparse frames
    (see module docstring): the quad's angle is a noisy prior, not a
    trustworthy one, so the window must be wide enough to correct it (Task
    23's brief default, 20 deg, comfortably covers the quad's typical error
    after the stance gate's own ~26 deg slack is accounted for) while still
    being narrow enough to search efficiently and not have to reproduce the
    full-sweep case.

    `init_center` given: points farther than 1.5*`side` away are dropped
    before fitting, so unrelated clutter elsewhere in a candidate's point
    cloud (e.g. from a generously-sized generator patch) can't bias the
    robust-extent center estimate away from the seeded region. When absent,
    all of `coords_2d` is used as-is.

    Returns None if too few points remain to fit.
    """
    coords_2d = np.asarray(coords_2d, dtype=np.float64)
    if len(coords_2d) < _MIN_POINTS:
        return None

    work = coords_2d
    if init_center is not None:
        init_center = np.asarray(init_center, dtype=np.float64)
        radius = _LOCALIZE_RADIUS_FACTOR * side
        near = np.linalg.norm(coords_2d - init_center, axis=1) < radius
        if near.sum() >= _MIN_POINTS:
            work = coords_2d[near]

    if len(work) < _MIN_POINTS:
        return None

    if init_theta is None:
        coarse_thetas = np.linspace(0.0, np.pi / 2.0, _COARSE_STEPS,
                                    endpoint=False)
        coarse_fit = _best_over(work, side, coarse_thetas)
        lo = coarse_fit.theta - np.radians(_COARSE_REFINE_WINDOW_DEG)
        hi = coarse_fit.theta + np.radians(_COARSE_REFINE_WINDOW_DEG)
    else:
        lo = init_theta - np.radians(theta_window_deg)
        hi = init_theta + np.radians(theta_window_deg)

    n_steps = max(9, int(round((hi - lo) / np.radians(_RESOLUTION_DEG))) + 1)
    fine_thetas = np.linspace(lo, hi, n_steps)
    return _best_over(work, side, fine_thetas)

File: rust/board-cluster-detector/src/square_fit.rs
//! Fixed-size square model fit: refine (or rescue) a candidate's pose by
//! fitting a KNOWN-side square's 3 DOF (center, theta) directly to its 2D
//! points, instead of trusting the 2D quad's (raw-point minAreaRect) angle.
//!
//! Port of `experiments/board-detection-2d/src/boarddet/square_fit.py`.
//!
//! Deliberately NOT filled-square ICP: an ICP-style fit that only pulls its
//! model's occupied interior onto matched points has zero residual gradient
//! whenever the model already encloses every point. The coverage residual
//! below charges for BOTH (a) points that fall outside the fixed-size square
//! and (b) perimeter band it doesn't reach, so an over-large or mis-rotated
//! enclosing box is still penalized and the search has a gradient to follow.

const MIN_POINTS: usize = 20;
// Trim this much off each end when reading the points' own extent per axis.
const EXTENT_PCTL: f64 = 2.0;
const N_BINS_PER_SIDE: usize = 10;
const BAND_FRAC: f64 = 0.06;
const COARSE_STEPS: usize = 37;
const COARSE_REFINE_WINDOW_DEG: f64 = 4.0;
const RESOLUTION_DEG: f64 = 0.25;
const LOCALIZE_RADIUS_FACTOR: f64 = 1.5;

/// Result of a fixed-side square fit.
#[derive(Debug, Clone, Copy)]
pub struct SquareFit {
    /// Plane coords, same frame as the input `coords`.
    pub center: [f64; 2],
    /// Radians; determined mod 90 deg (4-fold symmetry).
    pub theta: f64,
    /// Coverage residual (geometric + perimeter coverage), lower is better; ~0 is
    /// exact. This is the score the theta search minimises.
    pub residual: f64,
    /// The geometric half of `residual` alone: mean distance of points outside the
    /// square, over side. Unlike `residual` this stays meaningful when the sensor is
    /// too sparse to reach the perimeter, so it is what an acceptance gate should use
    /// for a small target at range (H-17).
    pub geometric_residual: f64,
    /// Cyclic order (not necessarily CCW).
    pub corners_2d: [[f64; 2]; 4],
}

/// `np.percentile(sorted_vals, p)` with numpy's default linear interpolation.
/// `sorted_vals` must already be sorted ascending and non-empty.
fn percentile(sorted_vals: &[f64], p: f64) -> f64 {
    let n = sorted_vals.len();
    if n == 1 {
        return sorted_vals[0];
    }
    let idx = p / 100.0 * (n as f64 - 1.0);
    let lo = idx.floor() as usize;
    let hi = idx.ceil() as usize;
    if lo == hi {
        return sorted_vals[lo];
    }
    let frac = idx - lo as f64;
    sorted_vals[lo] + frac * (sorted_vals[hi] - sorted_vals[lo])
}

/// `rel` is points in the square's own axis-aligned frame, already centered
/// on the candidate center (so the square occupies `[-side/2, side/2]^2`).
/// Lower is better; 0 is an exact, fully-covered fit.
/// The two halves of the coverage residual, reported separately.
///
/// `geometric` is the mean distance of points falling OUTSIDE the fixed-size square,
/// normalised by side. It measures how well the square models the points actually
/// observed, and is near zero for a real plate however sparsely it is sampled.
///
/// `coverage` is the fraction of perimeter bins holding no nearby point. It is what
/// gives the theta search a gradient -- an over-large or mis-rotated enclosing box is
/// penalised by it -- but it is NOT a measure of fit quality when the sensor cannot
/// reach the perimeter. A 600 mm plate at 7-8 m is crossed by about four VLP-32C rings,
/// so roughly half its perimeter bins are unfillable by construction (H-17).
///
/// `total` is `geometric + coverage`, the historical single number: it still selects
/// theta, so search behaviour is unchanged.
#[derive(Debug, Clone, Copy)]
pub struct ResidualTerms {
    pub geometric: f64,
    pub coverage: f64,
    pub total: f64,
}

fn coverage_residual(rel: &[[f64; 2]], side: f64) -> ResidualTerms {
    let half = side / 2.0;

    // Term 1: mean per-point distance outside the square.
    let n = rel.len() as f64;
    let mean_outside: f64 = rel
        .iter()
        .map(|p| {
            let dx = (p[0].abs() - half).max(0.0);
            let dy = (p[1].abs() - half).max(0.0);
            dx.hypot(dy)
        })
        .sum::<f64>()
        / n;

    // Term 2: fraction of the perimeter's binned bands with no nearby point.
    let corners_sq = [[half, half], [-half, half], [-half, -half], [half, -half]];
    let band = BAND_FRAC * side;
    let mut misses = 0usize;
    for i in 0..4 {
        let a = corners_sq[i];
        let b = corners_sq[(i + 1) % 4];
        let ab = [b[0] - a[0], b[1] - a[1]];
        let length = ab[0].hypot(ab[1]);
        let ab_hat = [ab[0] / length, ab[1] / length];

        let mut covered = [false; N_BINS_PER_SIDE];
        for p in rel {
            let r = [p[0] - a[0], p[1] - a[1]];
            let t = (r[0] * ab_hat[0] + r[1] * ab_hat[1]) / length;
            let perp = (r[0] * ab_hat[1] - r[1] * ab_hat[0]).abs();
            let near = perp < band && t > -0.02 && t < 1.02;
            if near {
                let bin_idx = ((t * N_BINS_PER_SIDE as f64) as i64)
                    .clamp(0, N_BINS_PER_SIDE as i64 - 1) as usize;
                covered[bin_idx] = true;
            }
        }
        misses += covered.iter().filter(|c| !**c).count();
    }
    let coverage_penalty = misses as f64 / (4 * N_BINS_PER_SIDE) as f64;

    let geometric = mean_outside / side;
    ResidualTerms {
        geometric,
        coverage: coverage_penalty,
        total: geometric + coverage_penalty,
    }
}

/// Closed-form center (robust per-axis extent midpoint) + coverage residual
/// for a fixed theta. Returns (residual, center_world, corners_world).
///
/// Matches numpy `p_sq = coords_2d @ rot` (row-vector @ matrix) with
/// `rot = [[c, -s], [s, c]]`, which works out to `p_sq = rot^T @ coords`,
/// i.e. `p_sq_x = x*c + y*s`, `p_sq_y = -x*s + y*c` (unrotate by theta).
/// The inverse transforms (`center_world = center_sq @ rot.T`,
/// `corners_world = (corners_sq + center_sq) @ rot.T`) work out to the
/// standard forward rotation `rot @ v`: `out_x = c*x - s*y`, `out_y = s*x + c*y`.
fn fit_at_theta(
    coords: &[[f64; 2]],
    side: f64,
    theta: f64,
) -> (ResidualTerms, [f64; 2], [[f64; 2]; 4]) {
    let (c, s) = (theta.cos(), theta.sin());

    let mut xs: Vec<f64> = Vec::with_capacity(coords.len());
    let mut ys: Vec<f64> = Vec::with_capacity(coords.len());
    let p_sq: Vec<[f64; 2]> = coords
        .iter()
        .map(|p| {
            let x = p[0] * c + p[1] * s;
            let y = -p[0] * s + p[1] * c;
            xs.push(x);
            ys.push(y);
            [x, y]
        })
        .collect();
    xs.sort_by(|a, b| a.partial_cmp(b).unwrap());
    ys.sort_by(|a, b| a.partial_cmp(b).unwrap());

    let lo = [percentile(&xs, EXTENT_PCTL), percentile(&ys, EXTENT_PCTL)];
    let hi = [
        percentile(&xs, 100.0 - EXTENT_PCTL),
        percentile(&ys, 100.0 - EXTENT_PCTL),
    ];
    let center_sq = [(lo[0] + hi[0]) / 2.0, (lo[1] + hi[1]) / 2.0];

    let rel: Vec<[f64; 2]> = p_sq
        .iter()
        .map(|p| [p[0] - center_sq[0], p[1] - center_sq[1]])
        .collect();

    let terms = coverage_residual(&rel, side);

    let half = side / 2.0;
    let corners_sq = [[half, half], [-half, half], [-half, -half], [half, -half]];

    // Forward rotation (standard rot @ v): out_x = c*x - s*y; out_y = s*x + c*y
    let rotate_fwd = |v: [f64; 2]| [c * v[0] - s * v[1], s * v[0] + c * v[1]];
    let center_world = rotate_fwd(center_sq);
    let corners_world = [0, 1, 2, 3].map(|i| {
        rotate_fwd([
            corners_sq[i][0] + center_sq[0],
            corners_sq[i][1] + center_sq[1],
        ])
    });

    (terms, center_world, corners_world)
}

fn best_over(coords: &[[f64; 2]], side: f64, thetas: &[f64]) -> SquareFit {
    // Theta is still selected by the FULL residual: the coverage term is what penalises
    // a mis-rotated or over-large enclosing square, so removing it from the search would
    // leave rotation nearly unconstrained. Only the reported geometric half is new.
    let mut best_residual = f64::INFINITY;
    let mut best: Option<(f64, f64, [f64; 2], [[f64; 2]; 4])> = None;
    for &theta in thetas {
        let (terms, center, corners) = fit_at_theta(coords, side, theta);
        if terms.total < best_residual {
            best_residual = terms.total;
            best = Some((theta, terms.geometric, center, corners));
        }
    }
    let (theta, geometric, center, corners) = best.expect("thetas must be non-empty");
    SquareFit {
        center,
        theta,
        residual: best_residual,
        geometric_residual: geometric,
        corners_2d: corners,
    }
}

/// Linspace matching numpy's `np.linspace(start, stop, n, endpoint)`.
fn linspace(start: f64, stop: f64, n: usize, endpoint: bool) -> Vec<f64> {
    if n == 0 {
        return vec![];
    }
    if n == 1 {
        return vec![start];
    }
    let divisor = if endpoint { n - 1 } else { n };
    let step = (stop - start) / divisor as f64;
    (0..n).map(|i| start + i as f64 * step).collect()
}

/// Fit a fixed-side-`side` square's center + theta to `coords`.
///
/// `init_theta = None`: no orientation prior at all — coarse full sweep over
/// `[0, 90 deg)` (4-fold symmetry), then a finer windowed refine around the
/// coarse winner.
///
/// `init_theta = Some`: search is a bounded window (+/- `theta_window_deg`)
/// around it instead of the full range.
///
/// `init_center = Some`: points farther than `1.5 * side` away are dropped
/// before fitting (only if at least `MIN_POINTS` remain), so unrelated
/// clutter can't bias the robust-extent center estimate.
///
/// Returns `None` if too few points remain to fit.
pub fn fit_fixed_square(
    coords: &[[f64; 2]],
    side: f64,
    init_center: Option<[f64; 2]>,
    init_theta: Option<f64>,
) -> Option<SquareFit> {
    fit_fixed_square_windowed(coords, side, init_center, init_theta, 20.0)
}

/// As [`fit_fixed_square`], with an explicit `theta_window_deg` for the
/// `init_theta = Some` case (defaults to 20 deg via [`fit_fixed_square`]).
pub fn fit_fixed_square_windowed(
    coords: &[[f64; 2]],
    side: f64,
    init_center: Option<[f64; 2]>,
    init_theta: Option<f64>,
    theta_window_deg: f64,
) -> Option<SquareFit> {
    if coords.len() < MIN_POINTS {
        return None;
    }

    let mut work: Vec<[f64; 2]> = coords.to_vec();
    if let Some(center) = init_center {
        let radius = LOCALIZE_RADIUS_FACTOR * side;
        let near: Vec<[f64; 2]> = coords
            .iter()
            .copied()
            .filter(|p| (p[0] - center[0]).hypot(p[1] - center[1]) < radius)
            .collect();
        if near.len() >= MIN_POINTS {
            work = near;
        }
    }

    if work.len() < MIN_POINTS {
        return None;
    }

    let (lo, hi) = if let Some(theta0) = init_theta {
        (
            theta0 - theta_window_deg.to_radians(),
            theta0 + theta_window_deg.to_radians(),
        )
    } else {
        let coarse_thetas = linspace(0.0, std::f64::consts::FRAC_PI_2, COARSE_STEPS, false);
        let coarse_fit = best_over(&work, side, &coarse_thetas);
        (
            coarse_fit.theta - COARSE_REFINE_WINDOW_DEG.to_radians(),
            coarse_fit.theta + COARSE_REFINE_WINDOW_DEG.to_radians(),
        )
    };

    let n_steps = (((hi - lo) / RESOLUTION_DEG.to_radians()).round() as i64 + 1).max(9) as usize;
    let fine_thetas = linspace(lo, hi, n_steps, true);
    Some(best_over(&work, side, &fine_thetas))
}
